### PR TITLE
Task2:Adding  instant search

### PR DESCRIPTION
**Pull request by snehaSaraJohn**
- Added instant search Features and its related Test cases
**Lint test Result Screenshot**
instant_search_lint_result

**Testcase result Screenshot**
instant_search_test

**E2e Result Screenshot**
e2e_test_result

**Working Video**
[screen-capture.webm](https://github.com/snehaSaraJohn/tmo-web-ui-developer-puzzle/assets/147800536/16a39c78-92ad-4202-9140-55bd414c2a3e)

### CODE REVIEW

File: web-puzzle-tmo-master/apps/okreads-e2e/src/integration/search-books.spec.ts [hunk 1]
```diff
@@ -5,13 +5,12 @@ describe('When: Use the search feature', () => {
...
-
-    cy.get('form').submit();
-
     cy.get('[data-testing="book-item"]').should('have.length.greaterThan', 1);
   });
 
-  xit('Then: I should see search results as I am typing', () => {
-    // TODO: Implement this test!
+  it('Then: I should see search results as I am typing', () => {
+    cy.get('input[type="search"]').type('angular');
```
Comment (SiddhuJajoo) [inline web-puzzle-tmo-master/apps/okreads-e2e/src/integration/search-books.spec.ts:12]: Above testcase and this is same. No need to add this, just update title of above one .

File: web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts [hunk 2]
```diff
@@ -53,8 +53,8 @@ describe('ProductsListComponent', () => {
       spyOn(store, 'dispatch');
     });
 
-    it('should be called when we have value for search term ', fakeAsync(() => {
-
+    it('searchBooks() should be called when we have value for search term and actual time spent after entering search term is equal to 500ms', fakeAsync(() => {
```
Comment (SiddhuJajoo) [inline web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts:56]: Keep it simple as `should dispatch searchBooks action to retrieve book details when search term is typed`

File: web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts [hunk 3]
```diff
@@ -64,14 +64,15 @@ describe('ProductsListComponent', () => {
...
-
-      component.searchForm.controls.term.setValue('');
+    it('No action should be dispatched when we have value for search term and actual time spent after entering search term is less than 500ms', fakeAsync(() => {
+      component.ngOnInit();
+      component.searchForm.controls.term.setValue('Java');
 
       tick(400);
 
       expect(store.dispatch).not.toHaveBeenCalled();
       discardPeriodicTasks();
```
Comment (SiddhuJajoo) [inline web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts:74]: Add testcase for `clearSearch` action

**Review by SiddhuJajoo — CHANGES_REQUESTED** [review]

File: web-puzzle-tmo-master/apps/okreads-e2e/src/integration/search-books.spec.ts (hunk 1 above)
Comment (snehaSaraJohn) [reply web-puzzle-tmo-master/apps/okreads-e2e/src/integration/search-books.spec.ts:6]: Resolved

File: web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts (hunk 2 above)
Comment (snehaSaraJohn) [reply web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.spec.ts]: Resolved

File: web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.ts [hunk 4]
```diff
@@ -38,6 +40,25 @@ export class BookSearchComponent {
...
+        }
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.unsubscribeObservable$.next();
+    this.unsubscribeObservable$.complete();
+  }
 
   addBookToReadingList(book: Book):void {
```
Comment (SiddhuJajoo) [inline web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.ts:63]: Are you using this searchBooks() anywhere? Is this really needed?

**Review by SiddhuJajoo — COMMENTED** [review]
All changes should be squashed

File: web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.ts (hunk 4 above)
Comment (snehaSaraJohn) [reply web-puzzle-tmo-master/libs/books/feature/src/lib/book-search/book-search.component.ts:63]: I used that inside function
 
  searchExample() :void{
    this.searchForm.controls.term.setValue('javascript');
    this.searchBooks();
  }

, Now i changed code such a way that we can eliminate this unnecessary function

**Pull request closed without merging**